File: src/main/java/com/integreety/yatspec/e2e/interceptor/LsdRestTemplateInterceptor.java
package com.integreety.yatspec.e2e.interceptor;

import com.integreety.yatspec.e2e.captor.http.RequestCaptor;
import com.integreety.yatspec.e2e.captor.http.ResponseCaptor;
import com.integreety.yatspec.e2e.captor.repository.model.InterceptedInteraction;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;

import java.io.IOException;

/**
 * Created to intercept rest template calls for Yatspec interactions.
 * Attempts to reset the input stream so that no data is lost on reading the response body
 */
@Slf4j
@RequiredArgsConstructor
public class LsdRestTemplateInterceptor implements ClientHttpRequestInterceptor {

    private final RequestCaptor requestCaptor;
    private final ResponseCaptor responseCaptor;

    @Override
    public ClientHttpResponse intercept(final HttpRequest request, final byte[] body, final ClientHttpRequestExecution execution) throws IOException {
        InterceptedInteraction interceptedInteraction = null;
        try {
            interceptedInteraction = requestCaptor.captureRequestInteraction(request, new String(body));
        } catch (final Throwable t) {
            log.error(t.getMessage(), t);
        }
        final ClientHttpResponse response = execution.execute(request, body);
        try {
            responseCaptor.captureResponseInteraction(request, response, interceptedInteraction.getTarget(), interceptedInteraction.getPath(), interceptedInteraction.getTraceId());
        } catch (final Throwable t) {
            log.error(t.getMessage(), t);
        }
        return response;
    }
}

File: core/src/main/kotlin/io/lsdconsulting/lsd/distributed/interceptor/interceptor/LsdRestTemplateInterceptor.kt
package io.lsdconsulting.lsd.distributed.interceptor.interceptor

import io.lsdconsulting.lsd.distributed.connector.model.InterceptedInteraction
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.RequestCaptor
import io.lsdconsulting.lsd.distributed.interceptor.captor.http.ResponseCaptor
import io.lsdconsulting.lsd.distributed.interceptor.config.log
import io.lsdconsulting.lsd.distributed.interceptor.interceptor.TimeHelper.getNow
import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Created to intercept rest template calls for LSD interactions.
 * Attempts to reset the input stream so that no data is lost on reading the response body
 */
class LsdRestTemplateInterceptor(
    private val requestCaptor: RequestCaptor,
    private val responseCaptor: ResponseCaptor,
) : ClientHttpRequestInterceptor {

    @Throws(IOException::class)
    override fun intercept(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution
    ): ClientHttpResponse {
        val startDateTime = getNow()

        val interceptedInteraction: InterceptedInteraction = try {
            requestCaptor.captureRequestInteraction(request, String(body))
        } catch (t: Throwable) {
            log().error(t.message, t)
            return execution.execute(request, body)
        }

        val response = execution.execute(request, body)

        val elapsedTime = TimeUnit.NANOSECONDS.toMillis(getNow() - startDateTime)

        try {
            responseCaptor.captureResponseInteraction(
                request,
                response,
                interceptedInteraction.target,
                interceptedInteraction.path,
                interceptedInteraction.traceId,
                elapsedTime
            )
        } catch (t: Throwable) {
            log().error(t.message, t)
        }
        return response
    }
}
